builder/Dockerfile

```diff
@@ -65,12 +60,63 @@ USER dw
 
 ENTRYPOINT [ "/bin/bash" ]
 
+########################################
+# The golang build stage
+########################################
+
+FROM ${builderbase} as golang
+
+WORKDIR /go
+
+ENV PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin:/buildroot/bin
+
+# Build the golang binaries
+ADD api api
+ADD cmd cmd
+ADD internal internal
+ADD pkg pkg
+ADD go.mod go.mod
+ADD go.sum go.sum
+RUN mkdir -p /go/bin && \
+	time go build -o /go/bin/ ./cmd/...
+
+########################################
+# The artifact build stage
+########################################
+
+FROM ${builderbase} as artifact-stage
+
+WORKDIR /buildroot/ambassador
+
+ENV PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin:/buildroot/bin
+
+# Copy in the golang binaries
+COPY --from=golang /go/bin/* /buildroot/bin/
+
+# Setup the python code
+ADD python python
+RUN time pip install --no-deps -e python
+
+# Run post-compile.sh
+ADD build-aux-local build-aux-local
+ADD post-compile.sh post-compile.sh
+RUN bash post-compile.sh
+
+# XXX: this will go away
+RUN mkdir -p /ambassador/sidecars && \
+    ln -s /buildroot/ambassador/python/post_update.py /ambassador/post_update.py && \
+    ln -s /buildroot/ambassador/python/watch_hook.py /ambassador/watch_hook.py && \
+    ln -s /buildroot/ambassador/python/kubewatch.py /ambassador/kubewatch.py
+
+# These will be extracted into the optimized image later
+ADD demo demo
+ADD docs/yaml/ambassador/ambassador-crds.yaml docs/yaml/ambassador/ambassador-crds.yaml
 
 ########################################
 # The optimized images
 ########################################
 
-FROM ${artifacts} as artifacts
+FROM artifact-stage as artifacts
```

Comment:
This is key. We now start from the artifact-stage, defined above, instead of $artifacts, which would have been a build container snapshot.